database/database.go

```diff
@@ -1,127 +1,585 @@
+
+		_ = count
+		for _, v := range wordsMapItem.occurenceSlice {
+			quoteSlice[v.enumid].Match += float32(v.count) / float32(wordsMapItem.totalOccurences)
+		}
+	}
+	return &quoteSlice
+}
+
+// StoreQuote stores a new quote
+// If the ID is not zero, StoreQuote will try to find the appropriate quote and overwrite it
+// If the ID is zero a new quote will be created
+func StoreQuote(q QuoteT) error {
+	var err error
+
+	// Verify connection to PostgreSQL database
+	err = postgresDatabase.Ping()
 	if err != nil {
-		log.Print("Cannot store teacher: ", err)
-		return err
+		postgresDatabase.Close()
+		return errors.New("At StoreQuote: " + err.Error())
+	}
+
+	if q.QuoteID == 0 {
+		// add quote to postgresDatabase
+		err = postgresDatabase.QueryRow(
+			`INSERT INTO quotes (TeacherID, Context, Text, Unixtime, Upvotes) VALUES ($1, $2, $3, $4, $5) RETURNING QuoteID`,
+			q.TeacherID, q.Context, q.Text, q.Unixtime, q.Upvotes).Scan(&q.QuoteID)
+		if err != nil {
+			return errors.New("At StoreQuote: " + err.Error())
+		}
+
+		// add quote to localDatabase
+		addQuoteToLocalDatabase(q)
+	} else {
+		// try to find corresponding entry postgresDatabase and overwrite it
+		var res sql.Result
+		res, err = postgresDatabase.Exec(
+			`UPDATE quotes SET TeacherID=$2, Context=$3, Text=$4, Unixtime=$5, Upvotes=$6 WHERE QuoteID=$1`,
+			q.QuoteID, q.TeacherID, q.Context, q.Text, q.Unixtime, q.Upvotes)
+		if err != nil {
+			return errors.New("At StoreQuote: " + err.Error())
+		}
+		if rowsAffected, _ := res.RowsAffected(); rowsAffected == 0 {
+			return errors.New("At StoreQuote: Could not find specified entry for overwrite")
+		}
+
+		// try to find corresponding entry in localDatabase and overwrite it
+		err = overwriteQuoteInLocalDatabase(q)
+		if err != nil {
+			return errors.New("At StoreQuote: " + err.Error())
+		}
 	}
 
 	return nil
 }
 
-// GetTeachers is a function to get teachers from the database
-func GetTeachers() ([]Teacher, error) {
+// DeleteQuote deletes the quote corresponding to the given ID from the database and the quotes slice
+// It will also modifiy the words map
+func DeleteQuote(ID int) {
+
+}
+
+/* -------------------------------------------------------------------------- */
+/*                         EXPORTED TEACHERS FUNCTIONS                        */
+/* -------------------------------------------------------------------------- */
+
+// GetTeachers returns a slice containing all teachers
+// The returned slice is not sorted
+func GetTeachers() *[]TeacherT {
+	localDatabase.mux.LockRead()
+	defer localDatabase.mux.UnlockRead()
+	teacherSlice := localDatabase.teacherSlice
+
+	return &teacherSlice
+}
 
-	rows, err := database.Query("SELECT id,name,note,title FROM teachers")
-	defer rows.Close()
+// StoreTeacher stores a new teacher
+// If the ID is not zero, StoreTeacher will try to find the corresponding teacher and overwrite it
+// If the ID is nil a new teacher will be created
+func StoreTeacher(t TeacherT) error {
```

Comment:
please split `StoreTeacher` into `CreateTeacher` and `UpdateTeacher`, see reasoning for splitting `StoreQuote` above